internal/services/v1/schema.go

```diff
@@ -105,72 +118,104 @@ func (ss *schemaServer) WriteSchema(ctx context.Context, in *v1.WriteSchemaReque
 			return nil, rewriteError(ctx, err)
 		}
 
-		newDefs.Add(nsdef.Name)
+		newObjectDefNames.Add(nsdef.Name)
 	}
 
+	// Start the transaction to update the schema.
 	_, err = ds.ReadWriteTx(ctx, func(ctx context.Context, rwt datastore.ReadWriteTransaction) error {
+		// Build a map of existing caveats to determine those being removed, if any.
+		existingCaveats, err := rwt.ListCaveats(ctx)
+		existingCaveatDefMap := make(map[string]*core.CaveatDefinition, len(existingCaveats))
+		existingCaveatDefNames := util.NewSet[string]()
+		if err != nil {
+			return err
+		}
+
+		for _, existingCaveat := range existingCaveats {
+			existingCaveatDefMap[existingCaveat.Name] = existingCaveat
+			existingCaveatDefNames.Add(existingCaveat.Name)
+		}
+
+		// For each caveat definition, perform a diff and ensure the changes will not result in type errors.
+		for _, caveatDef := range compiled.CaveatDefinitions {
+			if err := shared.SanityCheckCaveatChanges(ctx, rwt, caveatDef, existingCaveatDefMap); err != nil {
```

Comment:
Seems off we are passing `existingCaveatDefMap` as part of `SanityCheckCaveatChanges` when the caller could check existence of the key?

Reply:
It could, but then we lose the shared code with v1alpha1 WriteSchema